File: src/bot/db/client_messages.rs
//! structs and things for messages sent by the client to the server

// https://github.com/serde-rs/serde/issues/760
// workaround for now: https://github.com/serde-rs/serde/issues/760#issuecomment-499570311

use twilight_bot_utils::prelude::*;
use serde::Serialize;

pub trait ClientMessage: Serialize {}


#[derive(Serialize, Debug)]
pub struct SaveMessage {
	pub message: SaveMessageTagEnum,
	pub id: Id::<MessageMarker>,
	pub channel_id: Id::<ChannelMarker>,
	pub author_id: Id::<UserMarker>,
	pub content: String,
	pub attachment_urls: Vec<String>
}
#[derive(Serialize, Debug)]
pub enum SaveMessageTagEnum { #[serde(rename = "save_message")] Tag }
pub use SaveMessageTagEnum::Tag as SaveMessageTag;
impl ClientMessage for SaveMessage {}


#[derive(Serialize, Debug)]
pub struct GetMessage {
	pub message: GetMessageTagEnum,
	pub id: Id::<MessageMarker>
}
#[derive(Serialize, Debug)]
pub enum GetMessageTagEnum { #[serde(rename = "get_message")] Tag }
pub use GetMessageTagEnum::Tag as GetMessageTag;
impl ClientMessage for GetMessage {}


#[derive(Serialize, Debug)]
pub struct SaveUser {
	pub message: SaveUserTagEnum,
	pub id: Id::<UserMarker>,
	pub name: String,
	#[serde(with = "super::discriminator")]
	pub discriminator: u16
}
#[derive(Serialize, Debug)]
pub enum SaveUserTagEnum { #[serde(rename = "save_user_message")] Tag }
pub use SaveUserTagEnum::Tag as SaveUserTag;
impl ClientMessage for SaveUser {}


#[derive(Serialize, Debug)]
pub struct GetUser {
	pub message: GetUserTagEnum,
	pub id: Id::<UserMarker>
}
#[derive(Serialize, Debug)]
pub enum GetUserTagEnum { #[serde(rename = "get_user_message")] Tag }
pub use GetUserTagEnum::Tag as GetUserTag;
impl ClientMessage for GetUser {}

File: src/db-service/client-messages.ts
// structs and things for messages sent by the client to the server

import { z, literal, object, string, union } from "zod";
import { config_parser } from "./config";

export type SaveMessage = z.infer<typeof save_message_parser>;
export const save_message_parser = object({
	message: literal("save_message"),
	id: string(),
	channel_id: string(),
	author_id: string(),
	content: string(),
	attachment_urls: string().array()
}).transform(msg => ({ ...msg, _key: msg.id }));

export type GetMessage = z.infer<typeof get_message_parser>;
export const get_message_parser = object({
	message: literal("get_message"),
	id: string()
});

export type SaveUser = z.infer<typeof save_user_parser>;
export const save_user_parser = object({
	message: literal("save_user"),
	id: string(),
	name: string(),
	discriminator: string().refine(discriminator => /^\d{4}$/.test(discriminator)),
	avatar_url: string().url()
}).transform(msg => ({ ...msg, _key: msg.id }));

export type GetUser = z.infer<typeof get_user_parser>;
export const get_user_parser = object({
	message: literal("get_user"),
	id: string()
});

export type UpdateConfig = z.infer<typeof update_config_parser>;
export const update_config_parser = object({
	message: literal("update_config"),
	config: config_parser
});

export type GetConfig = z.infer<typeof get_config_parser>;
export const get_config_parser = object({
	message: literal("get_config"),
	id: string()
});

export type ClientMessages = z.infer<typeof client_messages>;
export const client_messages = union([
	save_message_parser,
	get_message_parser,
	save_user_parser,
	get_user_parser,
	update_config_parser,
	get_config_parser
]);
